Add APIM best practices (#514)
pagopa/dx: apps/website/docs/infrastructure/azure/apim/best-practices.md

---
sidebar_label: Best Practices
sidebar_position: 1
---

# API Management Best Practices

This documentation outlines few best practices for software engineers using
APIM.

The goal is to:

- Enhance security and confidence in API modifications by enabling testing in
  production
- Improve operational simplicity
- Increase developer autonomy

> ## Leveraging Versioning and Revisions for Safe Testing

Author: Agree, probably the title should be changed. Testing in production is explained later in `Testing APIs in Production Using Revisions`

APIM provides built-in capabilities for API versioning and revisions, offering a
structured way to manage API changes without disrupting active services.

Thus, leveraging on versioning and revisions is reccommended for:

- Testing in production
  - Modify policies and configurations safely before publishing changes
  - Conduct tests in production without impacting existing API consumers
  - Align versioning with downstream staging services (e.g. AppService and
    Function App staging slots)
- Keeping track of changes:
  - Maintain a clear changelog for tracking revisions
- Introducing breaking changes by maintaining previous API versions

### Versioning

Versioning allows multiple versions of an API Group to coexist. Clients can
request a specific version through:

- Path parameter (e.g. `/api/v1/resource`)
- Header parameter
- Query string parameter

If client does not specify a version, APIM fallbacks to the version chosen as

> default.

Reviewer: I won't mix developer tools (revisions) used to debug and other tools (versions) used to provide specific features to end users

> default.

Author: So are you suggesting to move revisions in debugging doc and to keep versions in best practices?

> default.

Reviewer: Yep, or deleting the version paragraph and include it sometime in the future into REST API best practice. For example, we should settle on just one method between header, path or query param to specify a version.

### Revisions

Revisions enable fine-grained control over API modifications. Each API version
can have multiple revisions with the following states:

- Current: The active revision used for API invocations
- Online: Available for invocation, but only with explicit revision selection
  (e.g. `/api/v1/resource;rev=2`).
- Offline: Inactive but can be promoted to online or current status

> ### Testing APIs in Production Using Revisions

Author: Why would you take off the section?

> ### Testing APIs in Production Using Revisions

Reviewer: Because imho the outer title suffices.

---

> sidebar_label: Best Practices

Reviewer: suggestion: I'd rather have a page titled something like 'Debug APIM Policies' with info about using revisions and VScode. Then another page about subscriptions and products titled 'Manage APIs access and policies'

sidebar_position: 1
---

# API Management Best Practices

This documentation outlines a few best practices for software engineers using
APIM.

The goal is to:

- Enhance security and confidence in API modifications by enabling testing in
  production
- Improve operational simplicity
- Increase developer autonomy

## Leveraging Versioning and Revisions for Safe Testing

APIM provides built-in capabilities for API versioning and revisions, offering a
structured way to manage API changes without disrupting active services.

> Thus, leveraging versioning and revisions is recommended for:

Reviewer: ```suggestion
Thus, leveraging revisions is recommended for:
```

Author: By changing the section title (as said above), I think we can keep this text as it is. This is a comprehensive list of unlocked potential by using these features. If you wish I can split them in two separate lists though

Reviewer: Version aren't used for testing or debugging, but to serve specific changes to end users (API clients). I won't mix the two concepts.

Author: Updated

- Testing in production
  - Modify policies and configurations safely before publishing changes
  - Conduct tests in production without impacting existing API consumers
  - Align versioning with downstream staging services (e.g. AppService and
    Function App staging slots)
- Keeping track of changes:
  - Maintain a clear changelog for tracking revisions
- Introducing breaking changes by maintaining previous API versions

> ### Versioning
> 
> Versioning allows multiple versions of an API Group to coexist. Clients can
> request a specific version through:
> 
> - Path parameter (e.g. `/api/v1/resource`)
> - Header parameter
> - Query string parameter
> 
> If the client does not specify a version, APIM falls back to the version chosen
> as default.

Reviewer: ```suggestion
```

> ### Revisions
> 
> Revisions enable fine-grained control over API modifications. Each API version

Reviewer: ```suggestion
### Leveraging Revisions for Debugging APIs and Policies

Each API version
```

can have multiple revisions with the following states:

- Current: The active revision used for API invocations
- Online: Available for invocation, but only with explicit revision selection
  (e.g. `/api/v1/resource;rev=2`).
- Offline: Inactive but can be promoted to online or current status

### Testing APIs in Production Using Revisions

1. Choose an existing API group
2. Define a new API version
3. Create revisions for incremental modifications (e.g. `rev 2`) and policy
   updates, and set as `Online`
4. Test the revision by invoking the API under test and specifying its revision
   name in the URL (e.g. `/api/v1/resource;rev=2`)
5. Finalize changes and promote a revision to current when ready
6. Document modifications using changelogs available through APIM

[VS Code can help you in debugging policies](debugging.md) via its built-in
debugger.

> ## Managing Subscriptions and Products

Reviewer: **thoughts**: I think our audience already understands how APIM subscriptions and products function; they need guidance on determining **granularity**, such as whether to create a product per _domain_ (ie. IO-COM), per API (ie. FIMS vs Firma con IO), or for an entire _product_ (ie. IO).

> 

Reviewer: ```suggestion
APIM **subscriptions** secure API access, while **products** group APIs with policies for specific audiences, used to manage access, quotas, and tailored API offerings.
```

> Let's illustrate how subscriptions and products function in APIM with an

Reviewer: ```suggestion
You can use them to to manage and secure API access, enforce usage limits, and deliver customized API offerings to different consumer groups, such as internal developers, partners, or public users.
```

always start with use cases (the _why_, instead of the _how_)

Author: The idea here is to explain the why through an example, to give the forma mentis instead of a dogma. Understanding the example you understand why things should be designed in this way
